### PR TITLE
Add warnings for big number of completions and parallelism

### LINKED ISSUE #118085: Indexed Jobs can break with high number of parallelism or completions
### What happened?

Currently, the maximum number of completions is MAX_INT32, around 2e9 and the maximum parallelism is 1e5.

When using these values, and assuming a very unlucky case, the field `.status.completedIndexes` can fill up the maximum size for objects in etcd (around 1.5Mi)

/sig apps
/sig scalability

### What did you expect to happen?

An Indexed Job with the maximum number of completions and parallelism to be correctly tracked.

### How can we reproduce it (as minimally and precisely as possible)?

```yaml
apiVersion: batch/v1
kind: Job
metadata:
  name: sample-job
spec:
  completionMode: Indexed
  completions: 2000000000
  parallelism: 100000
  template:
    spec:
      containers:
      - name: bad
        image: python
        command:
        - python3
        - -c
        - |
          import os
          import time
          id = int(os.environ.get("JOB_COMPLETION_INDEX"))
          if id > 1000000000 and id % 3 == 0:
            time.sleep(10000)
      restartPolicy: Never
```

In other words, every index multiple of 3 takes too long to terminate.

### Anything else we need to know?

In the example above, the number of characters needed by the `completedIndexes` field is:

`(10+1)*2*1e5`, roughly 2Mi, where:
- 10 is the number of characters for a number
- 1 for the separating character.
- 2 for the consecutive numbers
- 1e5 for the number of "holes" in the `completedIndexes`, based on the parallelism.

### Kubernetes version

1.21 and newer (when IndexedJob was introduced)

### Cloud provider

any

### OS version

<details>

```console
# On Linux:
$ cat /etc/os-release
# paste output here
$ uname -a
# paste output here

# On Windows:
C:\> wmic os get Caption, Version, BuildNumber, OSArchitecture
# paste output here
```

</details>

### Install tools

<details>

</details>

### Container runtime (CRI) and version (if applicable)

<details>

</details>

### Related plugins (CNI, CSI, ...) and versions (if applicable)

<details>

</details>

### PR DESCRIPTION
**Pull request by alculquicondor**
#### What type of PR is this?

/kind bug
/kind api-change

#### What this PR does / why we need it:

Add a warning for Jobs when `completions > 1e5` and `parallelism > 1e4`.

In this case, the size of the field `.status.completedIndexes` might exceed the storage limits of etcd, if a big number of pod indexes fail.

- If `parallelism <= 1e4`, then the field consumes at most `(10+1)*(10^4+10^4)=0.21Mi`.
- If `completions <= 1e5`, then the field consumes at most `(5+1)*10^5=0.572Mi`

These limits are well below the default etcd limits of ~1.5Mi

#### Which issue(s) this PR fixes:

Fixes #118085

#### Special notes for your reviewer:

This warning is in accordance with the proposal in https://github.com/kubernetes/enhancements/pull/3967.
I'll also submit a PR to update the documentation for Indexed Job.

#### Does this PR introduce a user-facing change?

```release-note
ACTION_REQUIRED
When an Indexed Job has a number of completions higher than 10^5 and parallelism higher than 10^4, and a big number of Indexes fail, Kubernetes might not be able to track the termination of the Job. Kubernetes now emits a warning, at Job creation, when the Job manifest exceeds both of these limits.
```

#### Additional documentation e.g., KEPs (Kubernetes Enhancement Proposals), usage docs, etc.:

```docs
- [Usage]: https://kubernetes.io/docs/concepts/workloads/controllers/job/#completion-mode
```

### CODE REVIEW

File: pkg/api/job/warnings_test.go [hunk 1]
```diff
@@ -0,0 +1,114 @@
...
+			wantWarningsCount: 1,
+		},
+	}
+	for name, tc := range cases {
+		t.Run(name, func(t *testing.T) {
+			_, ctx := ktesting.NewTestContext(t)
+			warnings := WarningsForJobSpec(ctx, nil, &tc.spec, nil)
+			if len(warnings) != tc.wantWarningsCount {
+				t.Errorf("Got %d warnings, want %d", len(warnings), tc.wantWarningsCount)
+				t.Logf("Warnings: %v", warnings)
```
Comment (mimowo) [inline pkg/api/job/warnings_test.go:106]: Is it left intentionally? I suppose it is a left-over of dev process. Maybe we could consider an additional field to match the warnings, but not requesting it.

File: pkg/api/job/warnings_test.go [hunk 2]
```diff
@@ -0,0 +1,114 @@
...
+		},
+		"valid Indexed medium completions medium parallelism": {
+			spec: batch.JobSpec{
+				CompletionMode: completionModePtr(batch.IndexedCompletion),
+				Completions:    pointer.Int32(100_000),
+				Parallelism:    pointer.Int32(100_000),
+				Template:       validPodTemplate,
+			},
+		},
+		"invalid Indexed high completions high parallelism": {
```
Comment (mimowo) [inline pkg/api/job/warnings_test.go:90]: I think it would be good to add test case with high completions and high parallelism when completionMode is NonIndexed, to show the validation is `if`ed

File: pkg/registry/batch/job/strategy_test.go [hunk 3]
```diff
@@ -892,6 +899,32 @@ func TestJobStrategy_WarningsOnCreate(t *testing.T) {
 				Spec: validSpec,
 			},
 		},
+		"high completions and parallelism": {
```
Comment (mimowo) [inline pkg/registry/batch/job/strategy_test.go:902]: I would like to have an analogous test for updates to ensure both entry paths to the new code are adjusted.

**Review by mimowo — COMMENTED** [review]
LGTM, some nits.

File: pkg/api/job/warnings_test.go (hunk 1 above)
Comment (alculquicondor) [reply pkg/api/job/warnings_test.go:106]: Yes, it's intentional. It could be useful for debugging if the test fails.

File: pkg/api/job/warnings.go [hunk 4]
```diff
@@ -0,0 +1,50 @@
...
+	parallelismSoftLimitForUnlimitedCompletions = 10_000
+)
+
+// WarningsForJobSpec produces warnings for fields in the JobSpec.
+// This function assumes that defaulting and validation has succceded.
+func WarningsForJobSpec(ctx context.Context, path *field.Path, spec, oldSpec *batch.JobSpec) []string {
+	var warnings []string
+	if spec.CompletionMode != nil && *spec.CompletionMode == batch.IndexedCompletion {
+		if *spec.Completions > completionsSoftLimit && *spec.Parallelism > parallelismSoftLimitForUnlimitedCompletions {
+			msg := "In Indexed Jobs with a number of completions higher than 10^5 and a parallelism higher than 10^4, Kubernetes might not be able to track completedIndexes when a big number of indexes fail"
```
Comment (mimowo) [inline pkg/api/job/warnings.go:40]: non-blocking nit: I'm wondering about adding something like "Consider splitting the Job into smaller ones"?

File: pkg/api/job/warnings_test.go [hunk 5]
```diff
@@ -0,0 +1,121 @@
...
+	"k8s.io/kubernetes/test/utils/ktesting"
+	"k8s.io/utils/pointer"
+)
+
+var (
+	validSelector = &metav1.LabelSelector{
+		MatchLabels: map[string]string{"a": "b"},
+	}
+
+	validPodTemplate = core.PodTemplateSpec{
```
Comment (mimowo) [inline pkg/api/job/warnings_test.go:34]: nit: I see declaring globally enables sharing, but on the other hand, if they are shared by many tests it might be not obvious if modifying them we don't change the intention of one of the tests that uses the shared objects. Thus, I would suggest declaring them scoped within a test,.

File: pkg/registry/batch/job/strategy_test.go [hunk 6]
```diff
@@ -749,6 +749,7 @@ func TestJobStrategy_WarningsOnUpdate(t *testing.T) {
 					Generation:      0,
 				},
 				Spec: batch.JobSpec{
+					CompletionMode: completionModePtr(batch.NonIndexedCompletion),
```
Comment (mimowo) [inline pkg/registry/batch/job/strategy_test.go:752]: nit: why adding this line? It seems unrelated.

**Review by mimowo — COMMENTED** [review]
Proposal to the warning message and two questions / suggestions regarding tests

File: pkg/api/job/warnings_test.go (hunk 5 above)
Comment (alculquicondor) [reply pkg/api/job/warnings_test.go:34]: The tests that use this don't care about the Pod template. Being verbose doesn't add much value.

File: pkg/registry/batch/job/strategy_test.go (hunk 6 above)
Comment (mimowo) [reply pkg/registry/batch/job/strategy_test.go]: I see, please do in the existing places. Adding in the new tests is good.

Comment (alculquicondor) [reply pkg/registry/batch/job/strategy_test.go]: removed.

File: pkg/api/job/warnings_test.go (hunk 5 above)
Comment (mimowo) [reply pkg/api/job/warnings_test.go:34]: I get it

File: pkg/api/job/warnings.go [hunk 7]
```diff
@@ -0,0 +1,50 @@
...
+	completionsSoftLimit                        = 100_000
+	parallelismSoftLimitForUnlimitedCompletions = 10_000
+)
+
+// WarningsForJobSpec produces warnings for fields in the JobSpec.
+// This function assumes that defaulting and validation has succceded.
+func WarningsForJobSpec(ctx context.Context, path *field.Path, spec, oldSpec *batch.JobSpec) []string {
+	var warnings []string
+	if spec.CompletionMode != nil && *spec.CompletionMode == batch.IndexedCompletion {
+		if *spec.Completions > completionsSoftLimit && *spec.Parallelism > parallelismSoftLimitForUnlimitedCompletions {
```
[the remark this answers is not in the document]
Comment (alculquicondor) [reply pkg/api/job/warnings.go:39]: API defaulting sets it to 1

[the remark this answers is not in the document]
Comment (deads2k) [reply pkg/api/job/warnings.go:39]: > API defaulting sets it to 1

Even so, given the pointer we should check nil-ness.  Defaulting could change in future versions for instance and it's easy to check now.

[the remark this answers is not in the document]
Comment (alculquicondor) [reply pkg/api/job/warnings.go]: I switched the implementation to `pointer.Int32Deref`

**Comment by mimowo** [comment]
The recent changes LGTM. Please take a look why the unit test failed, is it a flake?

**Comment by alculquicondor** [comment]
Yeah, I don't know why it's failing. I can't reproduce locally :(
